src/actions/actionHistory.tsx

```diff
@@ -0,0 +1,66 @@
+import { Action } from "./types";
+import React from "react";
+import { undo, redo } from "../components/icons";
+import { ToolButton } from "../components/ToolButton";
+import { t } from "../i18n";
+import { SceneHistory } from "../history";
+import { ExcalidrawElement } from "../element/types";
+import { AppState } from "../types";
+import { KEYS } from "../keys";
+
+const writeData = (
+  appState: AppState,
+  data: { elements: ExcalidrawElement[]; appState: AppState } | null,
+) => {
+  if (data !== null) {
+    return {
+      elements: data.elements,
+      appState: { ...appState, ...data.appState },
+    };
+  }
+  return {};
+};
+
+const testUndo = (shift: boolean) => (
+  event: KeyboardEvent,
+  appState: AppState,
+) =>
+  event[KEYS.META] &&
+  /z/i.test(event.key) &&
+  [
+    appState.multiElement,
+    appState.resizingElement,
+    appState.editingElement,
+    appState.draggingElement,
+  ].every(x => x === null) &&
```

Comment:
This seems to heavy a comparison to do on every keystroke (keep in mind that we need to run `keyTest` for every action, and if we do stuff like array loops, we'll soon be getting very slow  action handling).

But more importantly, to align with previous behavior we need to do this check inside the `perform`, otherwise this check won't be done for the button click.

Reply:
Also, it's not really a `keyTest`, semantically. And when we migrate to string-based key tests, this would not work, anyway.